src/dftbp/timedep/timeprop.F90

```diff
@@ -2097,22 +2094,18 @@ subroutine propagateRhoRealH(this, rhoOld, rho, H1, Sinv, step)
     T1R(:,:) = real(H1)
     T2R(:,:) = real(Sinv)
     call gemm(T3R,T2R,T1R)
+    T2R(:,:) = T3R
 
     ! calculate the first term products for the real and imaginary parts independently
     T1R(:,:) = real(rho)
-    T2R(:,:) = aimag(rho)
-    call gemm(T4R,T3R,T1R)
-    call gemm(T5R,T3R,T2R)
+    call gemm(T3R,T2R,T1R)
+
+    T1R(:,:) = aimag(rho)
+    call gemm(T4R,T2R,T1R)
 
     ! build the commutator combining the real and imaginary parts of the previous result
-    !$omp parallel do private(i,j)
-    do i=1,this%nOrbs
-      do j=1,this%nOrbs
-        rhoOld(i,j) = rhoOld(i,j) + cmplx(0, -step, dp) * (T4R(i,j) + imag * T5R(i,j)) &
-            + cmplx(0, step, dp) * conjg(T4R(j,i) + imag * T5R(j,i))
-      enddo
-    enddo
-    !$omp end parallel do
+    rhoOld(:,:) = rhoOld + cmplx(0, -step, dp) * (T3R + imag * T4R)&
+        & + cmplx(0, step, dp) * transpose(T3R - imag * T4R)
```

Comment:
If you remove the `parallel do` at least `workshare` this step.

Reply:
Suspect it would be better to just build with compiler options to use BLAS for matrix intrinsics.

Reply:
I don't see any BLAS operations in this step, other than maybe `zscale` or `zaxpy`, which as level 1 operations are unlikely to have large impact.

Reply:
Yep, level 1 only.

Curious how workshare will schedule the four matrix additions (simultaneously?)  if given enough processors.

Reply:
I guess this is hardly a time consuming step in the computation. You could run the timepropagation with Intel's VTune to see whether this routine appears in the hot spot / threading analysis.

Reply:
Yep (this routine does get called a lot, and effectively contains the only rank3 operations for fixed nuclei propagation, as overlap inversion would only happen once in that mode). Previous profiling suggested that for DFTB hamiltonians it's the interpolation of parameter tables that is the costly part.